src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1111,280 @@ private static string RegularExpressionFromFileSpec
-             * By definition, "**" must appear alone between directory slashes. If there is any remaining "**" then this is not
-             * a valid filespec.
-             */
-            // NOTE: this condition is evaluated left-to-right -- this is important because we want the length BEFORE stripping
-            // any "**"s remaining in the expression
-            if (matchFileExpression.Length > matchFileExpression.Replace(recursiveDirectoryMatch, null).Length)
+        /*
+         *  Replace characters with their regex equivalents.
+         */
+        private static void AppendRegularExpressionFromChar(StringBuilder regex, char ch)
+        {
+            if (ch == '*')
             {
-                isLegalFileSpec = false;
-                return String.Empty;
+                regex.Append(FileSpecRegexParts.AnyNonSeparator);
+            }
+            else if (ch == '?')
+            {
+                regex.Append(FileSpecRegexParts.SingleCharacter);
+            }
+            else if (FileUtilities.IsAnySlash(ch))
+            {
+                regex.Append(FileSpecRegexParts.DirSeparator);
+            }
+            else if (IsSpecialRegexCharacter(ch))
+            {
+                regex.Append('\\');
+                regex.Append(ch);
             }
+            else
+            {
+                regex.Append(ch);
+            }
+        }
 
-            /*
-             * Remaining call-outs not involving "**"
-             */
-            matchFileExpression.Replace("*.*", "<:anynonseparator:>");
-            matchFileExpression.Replace("*", "<:anynonseparator:>");
-            matchFileExpression.Replace("?", "<:singlecharacter:>");
+        private static bool IsSpecialRegexCharacter(char ch) =>
+            ch == '\\' || ch == '$' || ch == '(' || ch == ')' || ch == '+'
+            || ch == '.' || ch == '[' || ch == '^' || ch == '{' || ch == '|';
+
+        /*
+         * Iteratively skip three cases involving directory separators
+         * 
+         *  (1) \.\ -> \
+         *        This is an identity, so for example, these two are equivalent,
+         * 
+         *            dir1\.\dir2 == dir1\dir2
+         * 
+         *    (2) \\ -> \
+         *      Double directory separators are treated as a single directory separator,
+         *      so, for example, this is an identity:
+         * 
+         *          f:\dir1\\dir2 == f:\dir1\dir2
+         * 
+         *      The single exemption is for UNC path names, like this:
+         * 
+         *          \\server\share != \server\share
+         * 
+         *      This case is handled by isUncPath in
+         *      a prior step.
+         *
+         *  (3) **\** -> **
+         */
+        private static int IndexOfNextNonCollapsibleChar(string str, int startIndex)
+        {
+            int i = startIndex;
+            bool isFound = false;
+            bool isPrevRecursiveOperator = i > 2 && str[i - 3] == '*' && str[i - 2] == '*';
```

Comment:
Correct since ```\**\``` or ```**\``` will converted to the regex equivalent, and skipping over the consumed characters lands you after the ```\``` when calling this method with the offset and ```i + 1```. Although I just thought of possibly negating the ```(i + 1) and starting at the first wildcard.